src/compute/src/render/mod.rs

```diff
@@ -1120,6 +1142,40 @@ impl RenderTimestamp for Product<mz_repr::Timestamp, PointStamp<u64>> {
     }
 }
 
+/// A signal that can be awaited by operators to suspend them prior to startup.
+///
+/// Used to enforce a given configured hydration concurrency by only allowing that number of
+/// dataflows to perform hydration at the same time.
+///
+/// Note that currently only `persist_source` operators support suspension. Data can still enter a
+/// suspended dataflow through arrangement imports and ideally we would be able to suspend those as
+/// well.
+#[derive(Clone)]
+pub(super) struct StartSignal(
+    // The inner type is `Infallible` because no data is ever expected on this channel. Instead the
+    // signal is activated by dropping the corresponding `Sender`.
+    futures::future::Shared<oneshot::Receiver<Infallible>>,
+);
+
+impl StartSignal {
+    /// Create a new `StartSignal` and a corresponding token that activates the signal when
+    /// dropped.
+    pub fn new() -> (Self, Box<dyn Any>) {
+        let (tx, rx) = oneshot::channel::<Infallible>();
+        let token = Box::new(tx);
+        let signal = Self(rx.shared());
+        (signal, token)
+    }
+}
+
+impl Future for StartSignal {
+    type Output = ();
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> Poll<Self::Output> {
+        pin!(&mut self.0).poll(cx).map(|_| ())
```

Comment:
```suggestion
        Pin::new(&mut self.0).poll(cx).map(|_| ())
```

aren't these futures `Unpin`?

Reply:
Indeed, thanks! That was a leftover from trying to make the `watch`-based `StartSignal` work.